fl4health/model_bases/pca.py

```diff
@@ -0,0 +1,166 @@
+            Only used if self.low_rank is True. Defaults to 6.
+        """
+        super().__init__()
+        self.low_rank = low_rank
+        self.full_svd = full_svd
+        self.rank_estimation = rank_estimation
+        self.principal_components: Parameter
+        self.singular_values: Parameter
+        self.data_mean: Tensor
+
+    def forward(self, X: Tensor, center_data: bool) -> Tuple[Tensor, Tensor]:
+        """
+        Perform PCA on the data matrix X by computing its SVD.
+
+        Args:
+            X (Tensor): Data matrix.
+            center_data (bool): If true, then the data mean will be subtracted from all data points prior to
+            performing PCA.
+
+        Returns:
+            Tuple[Tensor, Tensor]: The principal components (i.e., left singular vectors) and their corresponding
+            singular values.
+        """
+        X_prime = self.prepare_data_forward(X, center_data=center_data)
+        if self.low_rank:
+            log(INFO, "Assuming data matrix is low rank, using low-rank PCA implementation.")
+            q = min(self.rank_estimation, X_prime.size(0), X_prime.size(1))
+            _, S, V = torch.pca_lowrank(X_prime, q=q, center=False)
+            principal_components = V
+        else:
+            if self.full_svd:
+                log(INFO, "Performing full SVD on data matrix.")
+            else:
+                log(INFO, "Performing reduced SVD on data matrix.")
+            _, S, Vh = torch.linalg.svd(X_prime, full_matrices=self.full_svd)
+            principal_components = Vh.T
+        singular_values = S
+        return principal_components, singular_values
+
+    def maybe_reshape(self, X: Tensor) -> Tensor:
+        """
+        Reshape input tensor X as needed so SVD can be computed.
+        """
+        if len(X.size()) == 2:
+            return torch.squeeze(X.float())
+        else:
+            dim0 = X.size(0)
+            return torch.squeeze(X.view(dim0, -1).float())
+
+    def set_data_mean(self, X: Tensor) -> None:
+        """
+        The primary purpose of this method is to store the mean of the
+        training data so it can be used to center validation/test data later,
+        if needed.
+        """
+        self.data_mean = torch.mean(X, dim=0)
+
+    def centre_data(self, X: Tensor) -> Tensor:
+        assert self.data_mean is not None
+        return X - self.data_mean
+
+    def prepare_data_forward(self, X: Tensor, center_data: bool) -> Tensor:
+        """
+        Prepare input data X for PCA by reshaping and centering it as needed.
+        """
+        X = self.maybe_reshape(X)
+        if center_data:
+            self.set_data_mean(X)
+            return self.centre_data(X)
+        else:
+            return X
+
+    def project_lower_dim(self, X: Tensor, k: Optional[int] = None) -> Tensor:
+        """
+        Project input data X onto the top k principal components.
+
+        Args:
+            X (Tensor): Input Data.
```

Comment:
This assumes that the data has been appropriately centered correct? We should probably support centering here as well (that is, applying the learned data mean to the input)?

Reply:
Good point!